h2o-core/src/main/java/water/FrameSizeMonitor.java

```diff
@@ -0,0 +1,211 @@
+package water;
+
+import water.fvec.Frame;
+import water.fvec.NewChunk;
+import water.parser.FVecParseWriter;
+import water.util.Log;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class FrameSizeMonitor extends MrFun<FrameSizeMonitor> {
+
+    private static final String LOG_LEVEL_PROP = "util.frameSizeMonitor.logLevel";
+    private static final String ENABLED_PROP = "util.frameSizeMonitor.enabled";
+    private static final String SAFE_COEF_PROP = "util.frameSizeMonitor.safetyCoefficient";
+    private static final String SAFE_FREE_MEM_DEFAULT_COEF = "0.2";
+
+    private static final int LOG_LEVEL;
+    private static final boolean ENABLED;
+    private static final float SAFE_FREE_MEM_COEF;
+    private static final int SLEEP_MS = 100;
+    private static final int MB = 1024 * 1024;
+    
+    private static final Map<Key<Job>, FrameSizeMonitor> registry = new HashMap<>();
+    
+    static {
+        LOG_LEVEL = Integer.parseInt(System.getProperty(LOG_LEVEL_PROP, String.valueOf(Log.DEBUG)));
+        ENABLED = H2O.getSysBoolProperty(ENABLED_PROP, false);
+        SAFE_FREE_MEM_COEF = Float.parseFloat(H2O.getSysProperty(SAFE_COEF_PROP, SAFE_FREE_MEM_DEFAULT_COEF));
+    }
+
+    private final Key<Job> jobKey;
+    private final Set<FVecParseWriter> writers = new HashSet<>();
+    private final long totalMemory = getTotalMemory();
+
+    private long committedMemory = 0;
+    
+    FrameSizeMonitor(Key<Job> jobKey) {
+        this.jobKey = jobKey;
+    }
+    
+    public static void get(Key<Job> jobKey, Consumer<FrameSizeMonitor> c) {
+        synchronized (registry) {
+            if (registry.containsKey(jobKey)) {
+                c.accept(registry.get(jobKey));
+            } else if (ENABLED) {
+                if (jobKey.get().stop_requested()) {
+                    // throw an exception to stop the parsing
+                    throw new IllegalStateException("Memory is running low. Forcefully terminating.");
+                } else {
+                    FrameSizeMonitor monitor = new FrameSizeMonitor(jobKey);
+                    registry.put(jobKey, monitor);
+                    H2O.submitTask(new LocalMR<>(monitor, 1));
+                    c.accept(monitor);
+                }
+            }
+        }
+    }
+    
+    private static void finish(Key<Job> jobKey) {
+        synchronized (registry) {
+            registry.remove(jobKey);
+        }
+    }
+
+    @Override
+    protected void map(int id) {
+        float nextProgress = 0.02f;
```

Comment:
Can we name the start and the progress step in some way please, so it's more explanatory ? Start is at `0.02f` and the step seems to be `0.01f` and then `0.1f`.